Restrictions on broadcaster withdrawals

yondonfu opened the pull request:
Current approach to restricting broadcaster deposit withdrawals:

The JobsManager keeps track of the number of active jobs for a broadcaster. A job can be in 3 states - `Active`, `Ended` and `Inactive`. When a job is created it becomes `Active`. When either a transcoder or broadcaster calls `endJob(jobId)` for a job, the job becomes `Ended` and an end block is set for the job. After a job's end block, it becomes `Inactive`. A broadcaster can only withdraw its deposit if it has 0 active jobs, and the latest end block for any of its ended jobs is less than or equal to the current block. A consequence of this is that a broadcaster must call `endJob` for all of its active jobs in order to withdraw its deposit which is not ideal from a UX/cost point of view, but let me know what you guys think.

As a solution to mitigate the possibility of a miner influencing the block hash used to determine which segments are challenged, we use the block hash of claimBlock + 1 which a miner can only influence if it wins the race to mine the block after the claimBlock.

Closes #76 #103

dob commented on contracts/jobs/JobsManager.sol:
```diff
@@ -198,8 +210,19 @@ contract JobsManager is ManagerProxyTarget, IVerifiable, IJobsManager {
+        uint256 endBlock = block.number.add(jobEndingPeriod);
+
         // Set end block for job
         job.endBlock = block.number.add(jobEndingPeriod);
```
Should we use the variable `endBlock` in this assignment since we just calculated it one line above?

dob commented on contracts/jobs/libraries/JobLib.sol:
```diff
@@ -51,13 +49,23 @@ library JobLib {
-        if (uint256(keccak256(_claimBlock, block.blockhash(_claimBlock), _segmentNumber)) % _verificationRate == 0) {
+        // Use block hash and block number of the block after a claim to determine if a segment
+        // should be verified
+        if (uint256(keccak256(_claimBlock + 1, block.blockhash(_claimBlock + 1), _segmentNumber)) % _verificationRate == 0) {
```
Is the fact that `block.blockhash` is only available for the most recent 256 blocks a problem here? Does it mean there's only a ~64 minute window for verification and slashing to take place before this code wouldn't work?

yondonfu replied:
Right...I forgot about that restriction. We probably cannot calculate the block hash in `shouldVerifySegment` and instead have to store the block hash at some point so that it can be used later

dob replied:
Ok, need to think about this more...since it seems like this limitation is a problem for us potentially.

yondonfu commented:
An alternative might be for there to either be a manually set or default end block for a job and after that block the job is considered inactive. When a broadcaster creates a job, it updates its own withdraw block to whatever is later, its current withdraw block or the end block of the new job. A broadcaster could optionally extend the length of a job via a function like `extendJob(uint256 _jobId, uint256 _blocks)` which updates the end block of the job. `withdraw` will just check if the broadcaster's withdraw block is less than or equal to the current block thereby avoiding the need for a broadcaster to call `endJob` for all of its active jobs. This would also mean jobs only need to have 2 states again `Active` or `Inactive`.

However, not directly related to withdrawals, I think there might still be potential issues with attacks related to the lifecycle of a job that I note here: https://github.com/livepeer/protocol/issues/46#issuecomment-342950346

yondonfu commented:
After discussing offline, the interim solution we are thinking of rolling with right now is as follows:

When a broadcaster creates a job, it can pass in an end block at which a job becomes inactive. Clients interacting with the contracts can set a default if a broadcaster does not manually set one (maybe 24 hours, maybe roundLength). As mentioned in the previous comment, the end block for a job also becomes the broadcaster's withdraw block if the end block > the broadcaster's withdraw block. `withdraw` checks if the current block >= the broadcaster's withdraw block.

yondonfu commented on contracts/jobs/JobsManager.sol:
```diff
@@ -238,6 +236,7 @@ contract JobsManager is ManagerProxyTarget, IVerifiable, IJobsManager {
                 segmentRange: _segmentRange,
                 claimRoot: _claimRoot,
                 claimBlock: block.number,
+                blockHash: block.blockhash(block.number - 1),
```
Since the previous code before this PR had the problem of using block.blockhash to reference the claim block when shouldVerifySegment is called it was possible for shouldVerifySegment to fail since block.blockhash only works for up to 256 blocks into the past.

Now we're storing a block hash when the claim is made to be used later. The block hash has to be in the past because the block hash of the current block during a claim is not known until the block is actually mined

dob replied:
Ok, this is definitely something we'll have to address before launch. We need to use a future value. I'm thinking an oracle service that submits blockhashes for block n would be ok, or restricting the verificationPeriod to 250 blocks, which would mean that the block + 1 hash would be available on chain, and it could be written down during the verify call.

yondonfu commented:
I updated the code to use the interim solution we discussed for withdrawal restrictions such that each newly created job is assigned an end block and the latest end block of a broadcaster's jobs becomes the broadcaster's withdraw block.

I also updated the segment challenge code to use a block hash stored during a claim. This does not solve the miner attack originally mentioned for this PR, but it does fix the block.blockhash issue that @dob pointed out that I describe in my inline comment. A fix for the miner attack can probably be pushed out of this PR and this PR can just focus on the withdrawal restrictions

dob commented:
Ok, looks good to me. I wonder if it will be a source of confusion for users that they have to set their own endblock, but since it can be passed in anyway I'm not sure how to provide the default without overloading the Job() function. Doing so would probably create more confusion. So let's leave it as a passed in value and set a default in the client. :ship:
